### PR TITLE
[VPlan] Compute cost for most opcodes in VPWidenRecipe (NFCI).

### PR DESCRIPTION
**Pull request by fhahn**
Implement VPWidenRecipe::computeCost for most cases (except 
UDiv,SDiv,URem,SRem which require additional logic).

Note that this specializes `::computeCost` instead of `::cost`, as
`VPRecipeBase::cost` is responsible for skipping cost-computations
for pre-computed recipes for now.

The most recent version of the VPlan-based cost model introduction 
has been committed on Jul 10 (b841e2eca3b5c8b) and we should
probably give it at least a week in case additional mismatches surface.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 1]
```diff
@@ -1137,6 +1137,80 @@ void VPWidenRecipe::execute(VPTransformState &State) {
...
+  case Instruction::FMul:
+  case Instruction::FDiv:
+  case Instruction::FRem:
+  case Instruction::Shl:
+  case Instruction::LShr:
+  case Instruction::AShr:
+  case Instruction::And:
+  case Instruction::Or:
+  case Instruction::Xor: {
+    VPValue *Op2 = getOperand(1);
```
Comment (rengolin) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1173]: nit: `RHS`?

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp [hunk 2]
```diff
@@ -1137,6 +1137,80 @@ void VPWidenRecipe::execute(VPTransformState &State) {
...
+  }
+  case Instruction::ICmp:
+  case Instruction::FCmp: {
+    Instruction *CtxI = dyn_cast_or_null<Instruction>(getUnderlyingValue());
+    Type *VectorTy = ToVectorTy(Ctx.Types.inferScalarType(getOperand(0)), VF);
+    return Ctx.TTI.getCmpSelInstrCost(Opcode, VectorTy, nullptr, getPredicate(),
+                                      CostKind, CtxI);
+  }
+  default:
+    llvm_unreachable("Unsupported opcode for instruction");
```
Comment (rengolin) [inline llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1210]: We don't want some kind of fall back?

File: llvm/lib/Transforms/Vectorize/VPlan.h [hunk 3]
```diff
@@ -862,7 +864,8 @@ class VPRecipeBase : public ilist_node_with_parent<VPRecipeBase, VPBasicBlock>,
 protected:
   /// Compute the cost of this recipe using the legacy cost model and the
   /// underlying instructions.
-  InstructionCost computeCost(ElementCount VF, VPCostContext &Ctx) const;
+  virtual InstructionCost computeCost(ElementCount VF,
+                                      VPCostContext &Ctx) const;
```
Comment (arcbbb) [inline llvm/lib/Transforms/Vectorize/VPlan.h:868]: nit: the comment can be refined as the legacy cost model can be overriden.

**Review by arcbbb — COMMENTED** [review]
lgtm

File: llvm/lib/Transforms/Vectorize/VPlan.h (hunk 3 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlan.h]: Adjusted the comment, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp]: Renamed, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp (hunk 2 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanRecipes.cpp:1213]: I think VPWidenRecipe recipes should only be created for the matched opcodes above and if there are any missed cases this should highlight them via a crash, similar to how we handle similar cases elsewhere.

**Comment by fhahn** [comment]
ping :)

**Review by arcbbb — APPROVED** [review]
Lgtm

**Comment by fhahn** [comment]
@arcbbb what do you think about porting over the implementations from https://github.com/llvm/llvm-project/pull/67647, in particular the memory-op cost ones?

**Review by ayalz — COMMENTED** [review]
Few post-commit nits.
Thanks for following-up after having split this from #92555

**Comment by arcbbb** [comment]
> @arcbbb what do you think about porting over the implementations from #67647, in particular the memory-op cost ones?

I just spoke with @ElvisWang123 about the recent changes on the VPlan. We’re planning to provide some initial implementations soon.

**Pull request merged**